rasa_core/featurizers.py

```diff
@@ -598,22 +608,26 @@ def prediction_states(self,
     def _remove_duplicate_states(
             trackers_as_states,  # type: List[List[Dict[Text, float]]]
             trackers_as_actions,  # type: List[List[Text]]
+            trackers_histories  #type: List[Dict[int, tuple[text,text]]]
     ):
         # type: (...) -> Tuple[List[List[Dict]], List[List[Dict]]]
         """Removes states that create equal featurizations.
 
         From multiple states that create equal featurizations
         we only need to keep one."""
 
-        hashed_featurizations = set()
+        hashed_featurizations = []
```

Comment:
this may be a problem in terms of memory and speed, as sets are significantly faster to lookup than  lists. it could however also be insignificant with respect to other points within the training mechanism

Reply:
it seems like this is a definite problem, will have a look at better ways of doing this today

Reply:
added a fix which uses a set and a dictionary to perform the list's functionality. slight memory cost and large gain in efficiency